packages/inferno/src/DOM/unmounting.ts

```diff
@@ -35,53 +35,82 @@ export function unmount(vNode) {
     }
 
     if (childFlags & ChildFlags.MultipleChildren) {
-      unmountAllChildren(children);
+      unmountAllChildren(children, animations);
     } else if (childFlags === ChildFlags.HasVNodeChildren) {
-      unmount(children as VNode);
+      unmount(children as VNode, animations);
     }
   } else if (children) {
     if (flags & VNodeFlags.ComponentClass) {
       if (isFunction(children.componentWillUnmount)) {
+        // TODO: Possible entrypoint
         children.componentWillUnmount();
       }
+
+      // If we have a willDisappear on this component, block children
+      let childAnimations = animations;
+      if (isFunction(children.willDisappear)) {
+        childAnimations = new AnimationQueues();
+        addDisappearAnimationHook(animations, children, children.$LI.dom);
+      }
+
       unmountRef(vNode.ref);
       children.$UN = true;
-      unmount(children.$LI);
+      unmount(children.$LI, childAnimations);
     } else if (flags & VNodeFlags.ComponentFunction) {
       ref = vNode.ref;
 
       if (!isNullOrUndef(ref) && isFunction(ref.onComponentWillUnmount)) {
+        // TODO: Possible entrypoint
         ref.onComponentWillUnmount(findDOMfromVNode(vNode, true) as Element, vNode.props || EMPTY_OBJ);
       }
 
-      unmount(children);
+      unmount(children, animations);
     } else if (flags & VNodeFlags.Portal) {
-      remove(children as VNode, vNode.ref);
+      // TODO: Possible entrypoint
+      remove(children as VNode, vNode.ref, animations);
     } else if (flags & VNodeFlags.Fragment) {
       if (vNode.childFlags & ChildFlags.MultipleChildren) {
-        unmountAllChildren(children);
+        // TODO: Possible entrypoint
+        unmountAllChildren(children, animations);
       }
     }
   }
 }
 
-export function unmountAllChildren(children: VNode[]) {
+export function unmountAllChildren(children: VNode[], animations: AnimationQueues) {
   for (let i = 0, len = children.length; i < len; ++i) {
-    unmount(children[i]);
+    unmount(children[i], animations);
   }
 }
 
-export function clearDOM(dom) {
-  // Optimization for clearing dom
-  dom.textContent = '';
+export function clearDOM(dom, animations: AnimationQueues) {
+  if (animations.willDisappear.length > 0) {
+    // Wait until animations are finished before removing actual dom nodes
+    // Be aware that the element could be removed by a later operation
+    callAllAnimationHooks(animations.willDisappear, () => { if (dom) dom.textContent = '' });
```

Comment:
function creation should be moved outside runtime of "clearDOM"

Reply:
Will fix

Reply:
@Havunen so I solved this by looking at how we optimised removeVNodeDOM. I don't fully grasp why this is better, but I get that it is a different way to create the anonymous callback function.

In other words, don't assume I know what I am doing and if I have misunderstood what you were going for, also take a look at `function removeVNodeDOM(...)` in common.ts. It uses the same pattern.

FIXED in 7cf58ef09